### PR TITLE
f-checkout@v0.6.0 – Test added for visibility on displayed fields.

### PR DESCRIPTION
**Pull request by JoshuaNg2332**
---

This PR is to add tests for f-checkout. This test confirms the existence of fields on the page.

---

## Browsers Tested

- [X] Chrome (latest)

### CODE REVIEW

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (matthewhardern) [inline packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: You can use `const` here as you never re-assign `areFieldsDisplayed`

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (matthewhardern) [inline packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: `isCheckoutFormFieldsDisplayed` does not currently exist in the imported utilities, and hence the reason for CI tests failing, did you forget to add it to git?

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (matthewhardern) [inline packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: Having logic inside tests In my personal opinion is not best practice. This logic should be moved to a function that itself can be tested. This prevents the test failing because of this logic.

Comment (SachaZvetelman) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: Normally we try not to "create new functionality or logic" inside a test. For example, you are creating a function, having conditionals and loops. The problem with this approach is that nothing is testing your test if it makes sense. Happy to explain more on a Hangouts, if you want!

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (SachaZvetelman) [inline packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: Nit: remove from the test once it's ready 😄

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (SachaZvetelman) [inline packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: I'm not entirely sure what we're checking here. Could you please clarify?
At the same time, remember the difference between `==` and `===`. 
In most scenarios, `if (!fields[i]) {` would suffice.

**Review by SachaZvetelman — COMMENTED** [review]
A few comments. Feel free to ping me if you have any questions, of course.
Make sure you set up ESLint on your VS Code. There are many issues that will be automatically fixed for you when saving the file (e.g. spacing), and it will warn you of important things such as "you can use `const` here".

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (kevinrodrigues) [inline packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: I think you could switch the entire contents of this test out for a simpler test. You would only need to test that that the individual fields exist, similar idea to what you are already doing. The [util](https://github.com/justeat/fozzie-components/blob/12d5aab2756c5c38dc186d150e5a745e4a8824a1/packages/f-checkout/test-utils/component-objects/f-checkout.component.js) contains all the data test IDs for you to be able to either loop or individually test them without the `areFieldsDisplayed` setup.

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (xander-marjoram) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: It looks like you're checking a list of boolean values to see if any of them are false, is that correct?

If so, you could use a function like [`filter`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) to check this rather than using a for-loop to iterate through the array.

```javascript
const fields = CheckoutComponent.isCheckoutFormFieldsDisplayed();
const hiddenFields = fields.filter(f => !f); // This removes any "truthy" elements
expect(hiddenFields.length).toBe(0);
```

Also, nit: I would rename `isCheckoutFormFieldsDisplayed` to something like `areFormFieldsDisplayed`. You could even move the bit where you filter it down into the method itself, and just return true or false inside there. 😄

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (kevinrodrigues) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: Looks to me that this is exported ok?

Comment (matthewhardern) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: On master `f-checkout.component` in the test-utils directory does not contain `isCheckoutFormFieldsDisplayed` unless I am mistaken? Also Circle CI error is because isCheckoutFormFieldsDisplayed is not a function. So assumed it was undefined, happy to be corrected though

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (matthewhardern) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: The above is a good shout, not sure on opinions here, but you could do a `it.each` passing in the data test id's of all the fields.

Comment (kevinrodrigues) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: `it.each` is something we would usually use here 👍

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (kevinrodrigues) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: Looks ok to me on master, would need to pull it to see what was happening, it does seem to already be used in this file further up too tho.

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (Festivejelly) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: @SachaZvetelman where would you put this logic then? It cant live in the page object model. All this is doing is asserting if fields are displayed. Its not changing the path the test would take, only the outcome of the assertion.

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (Festivejelly) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: Aka test.each i guess? https://jestjs.io/docs/en/api

Tbh we were struggling to find examples on how to do this because it.each isnt an available function.

Also the test i'ds dont seem to be in a collection so how would one pass each one in?

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (SachaZvetelman) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: @Festivejelly, my comment was about `console.log`, which shouldn't be there, since that is for dev purposes only. Or what part are you talking about exactly?

File: packages/f-checkout/test/specs/component/f-checkout.component.spec.js
Comment (matthewhardern) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: have a look at f-metadata `BrazeDispatcher.test.js` its got some good examples

Comment (xander-marjoram) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: @Festivejelly Yes `it` and `test` are interchangeable

Comment (Festivejelly) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: This is in Mocha. Your comments are relating to Jest afaik? We've added a new library to do a each on the it.

Comment (xander-marjoram) [reply packages/f-checkout/test/specs/component/f-checkout.component.spec.js]: I think we're picturing something along these lines
```javascript
it.each([
    'field1',
    'field2'
])('should display %s field', testId => {
    // Test that element with data-test-id=testId is displayed
});
```

File: package.json
Comment (siggerzz) [inline package.json]: You can remove the `^` from here 👍

File: wdio.conf.js
Comment (siggerzz) [inline wdio.conf.js]: This can be reverted

**Review by siggerzz — CHANGES_REQUESTED** [review]
Few minor comments, but otherwise looks good to me 🎉

**Review by siggerzz — DISMISSED** [review]

**Review by tommcclean — DISMISSED** [review]

**Comment by SachaZvetelman** [comment]
Only one question, and probably because I'm missing something, but why are we adding Mocha? :)

**Comment by ashleynolan** [comment]
Can you remove the JIRA ID's from your commit titles?

**Review by ashleynolan — CHANGES_REQUESTED** [review]
Just blocking this until the commit msgs are updated

**Comment by JoshuaNg2332** [comment]
> Just blocking this until the commit msgs are updated

@ashleynolan  Updated commit messages now.

**Comment by siggerzz** [comment]
@SachaZvetelman Mocha has always been part of the framework, we're just adding an additional mocha plugin. Mocha is the test framework that's use by WebDriverIO. (in the same way Jest is used for unit tests, or how we use Specflow in C#) 👍

**Comment by SachaZvetelman** [comment]
That's what I was missing: didn't know we were using Mocha with WebDriverIO. Now I do 😬

**Review by siggerzz — DISMISSED** [review]

File: packages/f-checkout/CHANGELOG.md
Comment (simonsandersje) [inline packages/f-checkout/CHANGELOG.md]: Change to today's date before merging

**Review by ashleynolan — DISMISSED** [review]

**Review by kevinrodrigues — DISMISSED** [review]

**Review by siggerzz — DISMISSED** [review]

**Review by simonsandersje — DISMISSED** [review]

File: packages/f-checkout/test-utils/component-objects/f-checkout.component.js
Comment (SachaZvetelman) [inline packages/f-checkout/test-utils/component-objects/f-checkout.component.js]: Sorry if I'm a bit late to the party, since you already have two approvals, but I feel this can be simplified.
For example, if you pass in the input as a parameter rather than the name, your function would pretty much be reduced to `return element.isDisplayed();`.

Comment (SachaZvetelman) [reply packages/f-checkout/test-utils/component-objects/f-checkout.component.js]: On a similar note, the function name should probably be singular in the end. You're checking that one given field is displayed, not all the checkout form fields are displayed, if it makes sense.

Comment (JoshuaNg2332) [reply packages/f-checkout/test-utils/component-objects/f-checkout.component.js]: I've had a chat with @siggerzz & @Festivejelly about this and we wanted to have a single test that checks these in one go rather than having individual functions that check visibility. If these were split into individual functions as part of the same test, one failure would stop the execution of the rest of the test.

Comment (SachaZvetelman) [reply packages/f-checkout/test-utils/component-objects/f-checkout.component.js]: Hmmm, I think we're talking about different things here, hehe. Let's have a chat after my training, before we merge this, please.

**Review by SachaZvetelman — CHANGES_REQUESTED** [review]
Just setting this "request changes" to avoid an accidental merge before we discuss those outstanding comments

**Review by siggerzz — DISMISSED** [review]

File: packages/f-checkout/test-utils/component-objects/f-checkout.component.js
Comment (SachaZvetelman) [inline packages/f-checkout/test-utils/component-objects/f-checkout.component.js]: Please remove commented code.

**Review by SachaZvetelman — DISMISSED** [review]
Approved with a minor comment.

File: packages/f-checkout/test-utils/component-objects/f-checkout.component.js
Comment (JoshuaNg2332) [reply packages/f-checkout/test-utils/component-objects/f-checkout.component.js]: Removed. 😄

**Review by SachaZvetelman — DISMISSED** [review]

**Review by siggerzz — DISMISSED** [review]

**Review by siggerzz — DISMISSED** [review]

File: CHANGELOG.md
Comment (ashleynolan) [inline CHANGELOG.md:14]: You can delete the latest line above by the looks of it and make this:

```
v2.9.0
------------------------------
*November 2, 2020*

### Added
- Added `mocha-each` to package.json.
- Added an entry to gitignore to ignore some local development tools for mocking http requests
```

**Review by ashleynolan — COMMENTED** [review]
Super small comment then LGTM

File: CHANGELOG.md
Comment (JoshuaNg2332) [reply CHANGELOG.md:14]: @ashleynolan I've had a look at this and thought that it follows the existing convention? The version number lives below the actual update and then the one above reading `Latest (to be added to next release)` needs to be there?
@siggerzz Do correct me if I'm wrong here?

Comment (ashleynolan) [reply CHANGELOG.md:14]: So the version number should be above each update. We use the `Latest` title to indicate that the package has been updated but didn't require a version bump, so those `CHANGELOG` entries get merged into the next release (so the `latest` changes there currently, merge into your release)

Comment (ashleynolan) [reply CHANGELOG.md:14]: Tbh, I'm not sure why there's a `latest` entry in the root of the mono repo anyway – it never gets released as such, so the version should just get bumped every change. So bit confusing – if you update it to be as I showed though that should resolve the issue.

**Review by siggerzz — APPROVED** [review]

**Review by ashleynolan — APPROVED** [review]

**Comment by ashleynolan** [comment]
Happy for this to be merged as long as you've checked the build is ok on CircleCI.

**Comment by siggerzz** [comment]
All good on CircleCI! - https://app.circleci.com/pipelines/github/justeat/fozzie-components/2118/workflows/6d8dc34c-fe01-4a93-8f6a-a47f7425ce9c

**Pull request merged**
